File: NASA_JulesKouatchou/Problem3/laplaceJacobi4.c
/*
 	This program solves Laplace's equation on the unit square.
        It relies on a fourth-order compact finite difference scheme and
 	uses the Jacobi iterative solver.
 
 	The BCs are Dirichlet with u(B) = 0 except u(x = 1, y) = 1.
 
 	In u(x_i,y_j) = u_ij = u[j][i], x[i] runs from xmin to xmax & y[j] runs
 	from ymin to ymax:
 		x = xmin + i*dx;
 		y = ymin + j*dy;
*/

#include <sys/types.h>
#include <time.h>
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#define MAX_SWEEPS	10000000

typedef struct {
	int N;
	double dx, xmin, xmax;
	double dy, ymin, ymax;
	double h;
	double omega, EPSILON;
} GRID;


double **allocate2d(int x, int y);
double timeStep4_Jacobi(GRID *grid, double *u[]);
void zero_matrix(GRID *grid, double *u[]);
void implement_BCs(GRID *grid, double *u[]);

int main( int argc, char *argv[] )
{
	double **u;
	GRID grid;
	double xmin = 0., xmax = 1.;
	double ymin = 0., ymax = 1.;
	double err, h, mu;
	double epsilon = 1.e-6;
	int N, i, sweep;
	char c[100];
        clock_t start, finish;

        // --------------------------------------------------
        // Get the number of grid points along each dimension
        // --------------------------------------------------
        if (argc == 2)
        {
           N = atoi(argv[1]);
        }
        else
        {
           printf("Usage: \n");
           printf("    %s dim \n", argv[0]);
           exit(-1);
        }

        start = clock();

        // ------------------------------
        // Allocate and initialize arrays
        // ------------------------------
	grid.N = N;
	grid.dx = (xmax-xmin)/N;
	grid.dy = (ymax-ymin)/N;
	h = grid.h = grid.dx;
	grid.xmin = xmin;
	grid.xmax = xmax;
	grid.ymin = ymin;
	grid.ymax = ymax;
	grid.EPSILON = epsilon*sqrt(h);

        u = allocate2d(N, N);

	zero_matrix(&grid, u);
	implement_BCs(&grid, u);
 
        err = 1000.0;

        // -----------------------------
        // Perform the Jacobi iterations
        // -----------------------------
        for (sweep = 0; sweep<MAX_SWEEPS && err > grid.EPSILON; sweep++) {
                  err = timeStep4_Jacobi(&grid, u);
        }

        finish = clock();
        printf("     Number of sweeps (%d): \n", sweep);
        printf("Time Jacobi Iteration (%d): %lf s\n", N, (double) (finish - start)/CLOCKS_PER_SEC);

	return 0;
}

double **allocate2d(int x, int y)
{
  int i;
  double **array = malloc(sizeof(double *) * x + sizeof(double) * x * y);
  for (i = 0; i < x; i++)
  {
    array[i] = ((double *)(array + x)) + y * i;
  }
  return array;
}

void zero_matrix(GRID *grid, double *u[])
{
	int N = grid->N, i, j;

	for (i = 0; i < N; i++) {
	    for (j = 0; j < N; j++) {
		u[i][j] = 0.;
            }
	}
}

void implement_BCs(GRID *grid, double *u[])
{
        int N = grid->N, i, j;
        double xmin = grid->xmin;
        double h = grid->h;
        double x;
        const double pi = 4. * atan(1.);

        for (j = 0; j < N; j++) {
                x = xmin + j*h;
                u[j][0] = sin(pi*x);
                u[j][N] = sin(pi*x)*exp(-pi);
        }
}

double timeStep4_Jacobi(GRID *grid, double *u[])
{
	int N = grid->N;
	double tmp, diff;
        int i, j;
        double err;

        err = 0.;

	for (i = 1; i < N-1; i++) {
		for (j = 1; j < N-1; j++) {
                        tmp = u[j][i];
                        u[j][i] = ( 4.*(u[j-1][i] + u[j+1][i] +
                                        u[j][i-1] + u[j][i+1]) + 
                                        u[j-1][i-1] + u[j+1][i+1] +
                                        u[j+1][i-1] + u[j-1][i+1])/20.;          
                        diff = u[j][i] - tmp;
                        err += diff*diff;
		}
	}
        return err;
}

File: NASA_JulesKouatchou/Problem3/laplaceJacobi4.py
#!/usr/bin/env python

################################################
# Numerical solution of the Laplace's equation.
#
#  u   + u   = 0
#   xx    yy
#
# Fourth-order compact Scheme
# Jacobi's iteration
################################################

import numpy
import sys
import time
import math


def slowTimeStep(u):
    """Takes a time step using straight forward Python loops."""
    n, m = u.shape

    err = 0.0
    for i in range(1, n-1):
        for j in range(1, m-1):
            tmp = u[i,j]
            u[i,j] = ((u[i-1, j] + u[i+1, j] +
                       u[i, j-1] + u[i, j+1])*4.0 +
                       u[i-1,j-1] + u[i-1,j+1] +
                       u[i+1,j-1] + u[i+1,j+1])/20.0

            diff = u[i,j] - tmp
            err += diff*diff

    return u,numpy.sqrt(err)

def numpyTimeStep(u):
    u_old=u.copy()
    u[1:-1, 1:-1] = ((u[0:-2, 1:-1] + u[2:, 1:-1] +
                      u[1:-1,0:-2] + u[1:-1, 2:])*4.0 +
                      u[0:-2, 0:-2] + u[0:-2, 2:] +
                      u[2:,0:-2] + u[2:, 2:])/20.0
    v = (u - u_old).flat
    return u,numpy.sqrt(numpy.dot(v,v))

# number of grid points
numPoints = int(sys.argv[1])

j=numpy.complex(0,1)
pi_c = 4.0*math.atan(1.0)

##################################
# Setting of the initial condition
##################################

u=numpy.zeros((numPoints,numPoints),dtype=float)

################################
# Setting the boundary condition
################################

x=numpy.r_[0.0:pi_c:numPoints*j]
u[0,:]=numpy.sin(x)
u[numPoints-1,:]=numpy.sin(x)*numpy.exp(-pi_c)

def regularSolver(u):
   iter =0
   err = 2
   while(iter <100000 and err>1e-6):
      (u,err)=slowTimeStep(u)
      iter+=1
   return (u,err,iter)

begTime1 = time.time()

(u,err,iter) = regularSolver(u)

endTime1 = time.time()
print " "
print "Regular Solver for Fourth-Order Scheme: ", numPoints
print "       Elapsed time: ", endTime1 - begTime1,"(s)"
print "       Number of iterations: ", iter
print "       Error:", err
#
# Vectorized Solver

v=numpy.zeros((numPoints,numPoints),dtype=float)

x=numpy.r_[0.0:pi_c:numPoints*j]
v[0,:]=numpy.sin(x)
v[numPoints-1,:]=numpy.sin(x)*numpy.exp(-pi_c)


def fastSolver(v):
   iter =0
   err = 2
   while(iter <100000 and err>1e-6):
      (v,err)=numpyTimeStep(v)
      iter+=1
   return (v,err,iter)

begTime2 = time.time()

(v,err,iter) = fastSolver(v)

endTime2 = time.time()
print " "
print "Vectorized Solver for Fourth-Order Scheme: ", numPoints
print "       Elapsed time: ", endTime2 - begTime2,"(s)"
print "       Number of iterations: ", iter
print "       Error:", err

